tests/functional/context/types/test_pure_types.py

```diff
@@ -0,0 +1,131 @@
+import pytest
+
+from vyper.context.types.value.address import AddressPureType
+from vyper.context.types.value.array_value import (
+    BytesArrayPureType,
+    StringPureType,
+)
+from vyper.context.types.value.boolean import BoolPureType
+from vyper.context.types.value.bytes_fixed import Bytes32PureType
+from vyper.context.types.value.numeric import (
+    DecimalPureType,
+    Int128PureType,
+    Uint256PureType,
+)
+from vyper.exceptions import (
+    InvalidLiteral,
+    OverflowException,
+    StructureException,
+    UnexpectedNodeType,
+    UnexpectedValue,
+)
+
+BASE_SOURCE = """
+"Add a docstring so the target test node doesn't become one."
+"""
+
+PURE_TYPES = {
+    AddressPureType: "address",
+    BoolPureType: "bool",
+    Bytes32PureType: "bytes32",
+    BytesArrayPureType: "bytes[1]",
+    DecimalPureType: "decimal",
+    Int128PureType: "int128",
+    StringPureType: "string[1]",
+    Uint256PureType: "uint256",
+}
+
+
+VALID_LITERALS = {
+    AddressPureType: ["0x6B175474E89094C44Da98b954EedeAC495271d0F"],
+    BoolPureType: ["True", "False"],
+    Bytes32PureType: ["0x6B175474E89094C44Da98b954EedeAC495271d0F4Da98b954EedeAC495271d0F"],
+    BytesArrayPureType: [
+        "b''",
+        "b'this is thirty three bytes long!!'",
+        r"b'\xbe\xef'",
+        "0x6B175474E89094C44Da98b954EedeAC495271d0F",
+        "0x6B175474E89094C44Da98b954EedeAC495271d0F4Da98b954EedeAC495271d0F",
+    ],
+    DecimalPureType: ["-1.666", "3.31337", "8008135.0", "1.2345678901"],
+    Int128PureType: ["-1", "0", "12", "42"],
+    StringPureType: ["''", "'hello'", "'this is thirty three chars long!!'", "'-42'"],
+    Uint256PureType: ["0", "12", "42"],
+}
+
+INVALID_LITERALS = [
+    "None",
+    "115792089237316195423570985008687907853269984665640564039457584007913129639936",
+    "115792089237316195423570985008687907853269984665640564039457584007913129639935.1",
+]
+INVALID_NODES = [
+    "self.{}",
+    "[{}]",
+    "({},)",
+    "{}()",
+    "{}(foo)",
+    "{}.foo",
+    "public({})",
+    "-{}",
+    "{}[1]",
+    "self.{}[1]",
+]
+
+
+@pytest.mark.parametrize("PureType", VALID_LITERALS.keys())
+@pytest.mark.parametrize(
+    "source", [x for v in VALID_LITERALS.values() for x in v] + INVALID_LITERALS
+)
+def test_from_literal(build_node, PureType, source):
```

Comment:
Lol, wtf is even happening here?

Reply:
All of these tests are like... way too clever lol

Reply:
added a comment